File: sdk/python/src/aclip/render_markdown.py
from __future__ import annotations

from typing import Any


def render_help_markdown(payload: dict[str, Any], tool_name: str) -> str:
    payload_type = payload["type"]
    if payload_type == "help_index":
        return _render_index(payload, tool_name)
    if payload_type == "help_command_group":
        return _render_group(payload, tool_name)
    if payload_type == "help_command":
        return _render_command(payload)
    raise ValueError(f"unsupported help payload type: {payload_type}")


def _render_index(payload: dict[str, Any], tool_name: str) -> str:
    lines = [
        f"# {tool_name}",
        "",
        payload["summary"],
        "",
    ]
    description = payload.get("description")
    if description:
        lines.extend([description, ""])
    if payload["command_groups"]:
        lines.extend(["## Command Groups", ""])
        for command_group in payload["command_groups"]:
            lines.append(f"- `{command_group['path']}`: {command_group['summary']}")
    if payload.get("commands"):
        if payload["command_groups"]:
            lines.append("")
        lines.extend(["## Commands", ""])
        for command in payload["commands"]:
            lines.append(f"- `{command['path']}`: {command['summary']}")
    lines.extend(
        [
            "",
            _render_next_line(tool_name),
            "",
        ]
    )
    return "\n".join(lines)


def _render_group(payload: dict[str, Any], tool_name: str) -> str:
    lines = [
        f"# {payload['path']}",
        "",
        payload["summary"],
        "",
    ]
    description = payload.get("description")
    if description:
        lines.extend([description, ""])
    if payload.get("command_groups"):
        lines.extend(["## Command Groups", ""])
        for command_group in payload["command_groups"]:
            lines.append(f"- `{command_group['path']}`: {command_group['summary']}")
        lines.append("")
    lines.extend(["## Commands", ""])
    for command in payload["commands"]:
        lines.append(f"- `{command['path']}`: {command['summary']}")
    lines.extend(
        [
            "",
            _render_next_line(tool_name),
            "",
        ]
    )
    return "\n".join(lines)


def _render_command(payload: dict[str, Any]) -> str:
    lines = [
        f"# {payload['path']}",
        "",
        payload["summary"],
        "",
    ]
    description = payload.get("description")
    if description:
        lines.extend([description, ""])

    lines.extend(
        [
            "## Usage",
            "",
            "```text",
            payload["usage"],
            "```",
            "",
            "## Arguments",
            "",
        ]
    )

    arguments = payload["arguments"]
    if not arguments:
        lines.append("- None")
    else:
        for argument in arguments:
            label = _render_argument_label(argument)
            state = "required" if argument["required"] else "optional"
            suffixes: list[str] = []
            if "default" in argument:
                suffixes.append(f"default `{argument['default']}`")
            if argument.get("multiple"):
                suffixes.append("repeatable")
            if "choices" in argument:
                choices = ", ".join(f"`{choice}`" for choice in argument["choices"])
                suffixes.append(f"choices {choices}")
            if "envVar" in argument:
                suffixes.append(f"env `{argument['envVar']}`")
            suffix = f", {', '.join(suffixes)}" if suffixes else ""
            lines.append(
                f"- `{label}` {state}{suffix}: {argument['description']}"
            )

    lines.extend(
        [
            "",
            "## Examples",
            "",
            "```text",
        ]
    )
    for example in payload["examples"]:
        lines.append(example)
    lines.extend(["```", ""])
    return "\n".join(lines)


def _render_argument_label(argument: dict[str, Any]) -> str:
    kind = argument["kind"]
    kind_label = _kind_token(kind)
    if "flag" in argument:
        return f"{argument['flag']} {kind_label}" if kind != "boolean" else argument["flag"]
    if "position" in argument:
        return f"<{argument['position']}:{kind}>"
    return argument["name"]


def _kind_token(kind: str) -> str:
    if kind == "integer":
        return "<integer>"
    if kind == "boolean":
        return ""
    return f"<{kind}>"


def _render_next_line(tool_name: str) -> str:
    if tool_name:
        return f"Next: run `{tool_name} <path> --help` for one command group or command shown above."
    return "Next: run `<tool> <path> --help` for one command group or command shown above."

File: sdk/typescript/src/renderMarkdown.ts
import type { HelpCommandGroupPayload, HelpCommandPayload, HelpIndexPayload, HelpPayload } from "./contracts.js";

export function renderHelpMarkdown(payload: HelpPayload, toolName: string): string {
  switch (payload.type) {
    case "help_index":
      return renderIndex(payload, toolName);
    case "help_command_group":
      return renderGroup(payload, toolName);
    case "help_command":
      return renderCommand(payload);
  }
}

function renderIndex(payload: HelpIndexPayload, toolName: string): string {
  const lines = [`# ${toolName}`, "", payload.summary, "", payload.description, ""];

  if (payload.command_groups.length) {
    lines.push("## Command Groups", "");
    for (const commandGroup of payload.command_groups) {
      lines.push(`- \`${commandGroup.path}\`: ${commandGroup.summary}`);
    }
    lines.push("");
  }

  if (payload.commands?.length) {
    lines.push("## Commands", "");
    for (const command of payload.commands) {
      lines.push(`- \`${command.path}\`: ${command.summary}`);
    }
    lines.push("");
  }

  lines.push(nextLine(toolName), "");
  return lines.join("\n");
}

function renderGroup(payload: HelpCommandGroupPayload, toolName: string): string {
  const lines = [`# ${payload.path}`, "", payload.summary, "", payload.description, ""];

  if (payload.command_groups?.length) {
    lines.push("## Command Groups", "");
    for (const commandGroup of payload.command_groups) {
      lines.push(`- \`${commandGroup.path}\`: ${commandGroup.summary}`);
    }
    lines.push("");
  }

  lines.push("## Commands", "");
  for (const command of payload.commands) {
    lines.push(`- \`${command.path}\`: ${command.summary}`);
  }
  lines.push("");
  return lines.join("\n");
}

function renderCommand(payload: HelpCommandPayload): string {
  const description = payload.description || payload.summary;
  const lines = [
    `# ${payload.path}`,
    "",
    description,
    "",
    "## Usage",
    "",
    "```text",
    payload.usage,
    "```",
    "",
    "## Arguments",
    ""
  ];

  if (!payload.arguments.length) {
    lines.push("- None");
  } else {
    for (const argument of payload.arguments) {
      lines.push(renderArgument(argument));
    }
  }

  lines.push("", "## Examples", "", "```text");
  for (const example of payload.examples) {
    lines.push(example);
  }
  lines.push("```", "");
  return lines.join("\n");
}

function renderArgument(argument: Record<string, unknown>): string {
  const kind = String(argument.kind);
  const label = Array.isArray(argument.flags)
    ? kind === "boolean"
      ? argument.flags.map((flag) => String(flag)).join(", ")
      : `${argument.flags.map((flag) => String(flag)).join(", ")} ${kindToken(kind)}`
    : "flag" in argument
      ? kind === "boolean"
        ? String(argument.flag)
        : `${String(argument.flag)} ${kindToken(kind)}`
      : `<${String(argument.position)}:${kind}>`;
  const state = argument.required ? "required" : "optional";
  const suffixes: string[] = [];

  if ("default" in argument) {
    suffixes.push(`default \`${String(argument.default)}\``);
  }
  if (argument.multiple) {
    suffixes.push("repeatable");
  }
  if (Array.isArray(argument.choices)) {
    suffixes.push(`choices ${argument.choices.map((choice) => `\`${String(choice)}\``).join(", ")}`);
  }
  if (typeof argument.envVar === "string") {
    suffixes.push(`env \`${argument.envVar}\``);
  }

  const suffix = suffixes.length ? `, ${suffixes.join(", ")}` : "";
  return `- \`${label}\` ${state}${suffix}: ${String(argument.description)}`;
}

function kindToken(kind: string): string {
  if (kind === "integer") {
    return "<integer>";
  }
  if (kind === "boolean") {
    return "";
  }
  return `<${kind}>`;
}

function nextLine(toolName: string): string {
  return `Next: run \`${toolName} <path> --help\` for one command group or command shown above.`;
}
